lib/sqlalchemy/sql/functions.py

```diff
@@ -49,7 +55,48 @@ def register_function(identifier, fn, package="_default"):
 
     """
     reg = _registry[package]
-    reg[identifier] = fn
+    case_sensitive_reg = _case_sensitive_registry[package]
+    raw_identifier = identifier
+    identifier = identifier.lower()
+
+    # Check if a function with the same lowercase identifier is registered.
+    if identifier in reg:
+        if raw_identifier in case_sensitive_reg[identifier]:
+            warnings.warn(
+                "The GenericFunction '{}' is already registered and "
+                "is going to be overriden.".format(identifier),
+                sa_exc.SAWarning)
+            reg[identifier] = fn
+        else:
+            # If a function with the same lowercase identifier is registered,
+            # then these 2 functions are considered as case-sensitive.
+            # Note: This case should raise an error in a later release.
+            reg.pop(identifier)
+
+    # Check if a function with different letter case identifier is registered.
+    elif identifier in case_sensitive_reg:
+        # Note: This case will be removed in a later release.
+        if (
+            raw_identifier not in case_sensitive_reg[identifier]
+        ):
+            util.warn_deprecated(
+                "GenericFunction(s) {} are already registered with "
+                "different letter cases and might interact with {}.".format(
```

Comment:
**mike bayer** (zzzeek) wrote:

..."GenericFunction objects will be fully case-insensitive in a future release."

i've been trying to make all the warnings hit home w/ the users :)